net/respondd/src/respondd.c

```diff
@@ -41,150 +48,381 @@
+
+	return ret;
+}
 
-  read_bytes = recvfrom(sock, buffer, sizeof(buffer), 0, (struct sockaddr *)&addr, &addrlen);
+static struct json_object * single_request(char *type) {
+	ENTRY key = {
+		.key = type,
+		.data = NULL,
+	};
+	ENTRY *entry = hsearch(key, FIND);
+	if (!entry)
+		return NULL;
 
-  if (read_bytes < 0) {
-    perror("recvfrom failed");
-    exit(EXIT_FAILURE);
-  }
+	struct request_type *r = entry->data;
 
-  clock_gettime(CLOCK_MONOTONIC, &tp);
+	if (r->cache_time && now < r->cache_timeout)
+		return json_object_get(r->cache);
 
-  lua_pushlstring(L, buffer, read_bytes);
-  lua_pushnumber(L, tp.tv_sec + 1.0e-9*tp.tv_nsec);
+	fputs("Generating\n", stderr);
 
-  if (lua_pcall(L, 2, 1, 0)) {
-    fprintf(stderr, "%s\n", lua_tostring(L, -1));
-  }
-  else if (lua_isstring(L, -1)) {
-    size_t msg_length;
-    const char *msg = lua_tolstring(L, -1, &msg_length);
+	struct json_object *ret = eval_providers(r->providers);
 
-    if (sendto(sock, msg, msg_length, 0, (struct sockaddr *)&addr, addrlen) < 0)
-      perror("sendto failed");
-  }
+	if (r->cache_time) {
+		if (r->cache)
+			json_object_put(r->cache);
 
-  lua_pop(L, 1);
+		r->cache = json_object_get(ret);
+		r->cache_timeout = now + r->cache_time;
+	}
+
+	return ret;
 }
 
+static struct json_object * multi_request(char *types) {
+	struct json_object *ret = json_object_new_object();
+	char *type, *saveptr;
+
+	for (type = strtok_r(types, " ", &saveptr); type; type = strtok_r(NULL, " ", &saveptr)) {
+		struct json_object *sub = single_request(type);
+		if (sub)
+			json_object_object_add(ret, type, sub);
+	}
+
+	return ret;
+}
+
+static struct json_object * handle_request(char *request, bool *compress) {
+	update_time();
+
+	if (!strncmp(request, "GET ", 4)) {
+		*compress = true;
+		return multi_request(request+4);
+	}
+	else {
+		*compress = false;
+		return single_request(request);
+	}
+}
+
+
+static void serve(int sock) {
+	char input[256];
+	const char *output = NULL;
+	ssize_t input_bytes, output_bytes;
+	struct sockaddr_in6 addr;
+	socklen_t addrlen = sizeof(addr);
+	bool compress;
+
+	input_bytes = recvfrom(sock, input, sizeof(input)-1, 0, (struct sockaddr *)&addr, &addrlen);
+
+	if (input_bytes < 0) {
+		perror("recvfrom failed");
+		exit(EXIT_FAILURE);
+	}
+
+	input[input_bytes] = 0;
+
+	struct json_object *result = handle_request(input, &compress);
+	if (!result)
+		return;
+
+	const char *str = json_object_to_json_string_ext(result, JSON_C_TO_STRING_PLAIN);
+
+	if (compress) {
+		size_t str_bytes = strlen(str);
+
+		mz_ulong compressed_bytes = mz_compressBound(str_bytes);
+		unsigned char *compressed = alloca(compressed_bytes);
+
+		if (!mz_compress(compressed, &compressed_bytes, (const unsigned char *)str, str_bytes)) {
+			output = (const char*)compressed;
+			output_bytes = compressed_bytes;
+		}
+	}
+	else {
+		output = str;
+		output_bytes = strlen(str);
+	}
+
+	if (output) {
+		if (sendto(sock, output, output_bytes, 0, (struct sockaddr *)&addr, addrlen) < 0)
```

Comment:
Isn't `< output_bytes` more honest? As we do not draw any consequences from the failed call, we might just as well check if the whole message was sent.